src/PreflightTest.ts

```diff
@@ -37,176 +77,673 @@ export interface PreflightTest {
+        throw constructInvalidValueError(
+          PreflightTest.Event.QualityWarning,
+          'element-in-currentWarnings',
+          'string',
+          typeof w
+        );
+      }
+    });
+
+    const previousWarnings =
+      nativeEvent[
+        Constants.PreflightTestQualityWarningEventKeyPreviousWarnings
+      ];
+    if (!Array.isArray(previousWarnings)) {
+      throw constructInvalidValueError(
+        PreflightTest.Event.QualityWarning,
+        'previousWarnings',
+        'array',
+        typeof previousWarnings
+      );
+    }
+    previousWarnings.forEach((w) => {
+      if (typeof w !== 'string') {
+        throw constructInvalidValueError(
+          PreflightTest.Event.QualityWarning,
+          'element-in-previousWarnings',
+          'string',
+          typeof w
+        );
+      }
+    });
+
+    this.emit(
+      PreflightTest.Event.QualityWarning,
+      currentWarnings as Call.QualityWarning[],
+      previousWarnings as Call.QualityWarning[]
+    );
+  };
 
-  private _handleSampleEvent = (_nativeEvent: NativePreflightTestEvent) => {
-    // TODO
-  }
+  /**
+   * Handle sample event.
+   */
+  private _handleSampleEvent = (
+    nativeEvent: PreflightTestType.NativeEventSample
+  ) => {
+    const sample = nativeEvent[Constants.PreflightTestSampleEventKeySample];
+    if (typeof sample !== 'string') {
+      throw constructInvalidValueError(
+        PreflightTest.Event.Sample,
+        'sample',
+        'string',
+        typeof sample
+      );
+    }
+
+    const parsedSample = JSON.parse(sample);
+
+    this.emit(PreflightTest.Event.Sample, parsedSample);
+  };
 
+  /**
+   * Get the CallSid of the underlying Call in the PreflightTest.
+   *
+   * @returns
+   * Promise that
+   * - Resolves with a string representing the CallSid.
+   * - Rejects if the native layer could not find the CallSid for this
+   *   PreflightTest object.
+   */
   public async getCallSid(): Promise<string> {
-    // TODO
     return NativeModule.preflightTest_getCallSid(this._uuid);
   }
 
+  /**
+   * Get the end time of the PreflightTest.
+   *
+   * @returns
+   * A Promise that
+   * - Resolves with `number` if the PreflightTest has ended.
+   * - Resolves with `undefined` if PreflightTest has not ended.
+   * - Rejects if the native layer encountered an error.
+   */
   public async getEndTime(): Promise<number> {
-    // TODO
     return NativeModule.preflightTest_getEndTime(this._uuid);
   }
 
+  /**
+   * Get the latest stats sample generated by the PreflightTest.
+   *
+   * @returns
+   * A Promise that
+   * - Resolves with a JSON-serialized `string` representing the last sample
+   *   generated by the PreflightTest.
+   * - Resolves with `undefined` if there is no previously generated sample.
+   * - Rejects if the native layer encountered an error.
+   */
   public async getLatestSample(): Promise<string> {
-    // TODO
     return NativeModule.preflightTest_getLatestSample(this._uuid);
   }
```

Comment:
Update the return type to PreflightTest.StatsSample.
Do you also need to parse the return value from preflightTest_getLatestSample?

Reply:
Yeah we will. I will add that kind of logic when I begin to implement the native layers.